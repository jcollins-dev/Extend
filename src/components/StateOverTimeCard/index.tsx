// 3rd party libs
import React, { useEffect, useMemo, useState } from 'react';
import styled from 'styled-components';
import { useTranslation } from 'react-i18next';

// Components
import { ActionButton, Card, GraphLegend, StateOverTimeChart, Typography } from 'components';
import {
  BarPeriod,
  Padding,
  Row,
  ScatterChartData,
  TrackingLine
} from 'components/StateOverTimeChart';

// Types
import { StatePeriod, ProteinMachineState, ProteinMachineCategoryStates } from 'types/protein';
import { ZoomObjectTuples } from 'types';
import { BrushProps } from 'types/graph';

export interface NestedRow {
  id: number;
  label: string;
  parentProperty: string;
  data: StatePeriod[];
  isLabel?: boolean;
  children?: NestedRow[];
}

type Props = {
  nestedData: NestedRow[];
  stateColors: Record<string, string>;
  title: string | React.ReactNode;
  subHeadingComponent?: React.ReactNode;
  // Flag to sync zooming with other graphs
  sync?: boolean;
  brush?: BrushProps;
  trackingLines?: TrackingLine[];
  scatterChartData?: { title?: string; data: ScatterChartData[] };
  barSpacing?: number;
  barCornerRadius?: number;
  intervalSpacing?: number;
  tickLabelPadding?: string | number;
  chartPadding?: Padding;
  hideStateCodes?: boolean;
  hideSubStepIds?: boolean;
  borderColor?: string;
  isDefaultExpanded?: boolean;
};

// Map state codes to state names
interface CodeNameMap {
  [code: string]: {
    stateName: string;
    stateCode: string | ProteinMachineState | ProteinMachineCategoryStates;
    uniqueStateCode: string;
  };
}

const graphPadding = {
  top: 50,
  right: 30,
  bottom: 5,
  left: 150
};

const Container = styled(Card)`
  display: flex;
`;

const Legends = styled.div`
  width: 11.5625rem;
  padding: 1rem;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  background-color: ${({ theme }) => theme.colors.lightGrey1};
  border-right: 1px solid ${({ theme }) => theme.colors.lightGrey2};
`;

const LegendRow = styled.div`
  margin-bottom: 0.5rem;
`;

const ResetZoom = styled.div`
  margin-bottom: 0.5rem;
`;

/**
 * Create an array of BarPeriods to populate a row in the chart, filtering out
 * any bars that are not active (i.e. the legend has been unselected)
 */
const createBars = (
  data: StatePeriod[],
  colors: Record<string, string>,
  activeState: Record<string, boolean>,
  overrideCode?: string
) => {
  return data.reduce((prev, current) => {
    const { stateCode, recipeName, stateName, startTimestamp, endTimestamp } = current;
    if (activeState[stateCode]) {
      const startTime = new Date(startTimestamp);
      const endTime = new Date(endTimestamp);
      prev.push({
        state: overrideCode ? overrideCode : (stateCode as string),
        color: colors[stateCode],
        startTime,
        endTime,
        toolTipData: {
          title: recipeName ? recipeName : stateName,
          startTime,
          endTime
        }
      });
    }
    return prev;
  }, [] as BarPeriod[]);
};

// Initial state for expanded rows (all collapsed by default)
const initExpandedState = (nestedData: NestedRow[], isDefaultExpanded = false) => {
  const state: Record<string, boolean> = {};
  nestedData.forEach((parentRow) => {
    state[parentRow.parentProperty] = isDefaultExpanded;
  });
  return state;
};

// Initial state for active states (i.e. legends that are currently selected)
// All states (rows) are shown by default
const initActiveState = (nestedData: NestedRow[]): Record<string, boolean> => {
  const active: Record<string, boolean> = {};
  nestedData.forEach((parentRow) => {
    parentRow.data.forEach(({ stateCode }) => {
      active[stateCode] = true;
    });

    parentRow.children?.forEach((childRow) => {
      childRow.data.forEach(({ stateCode }) => {
        active[stateCode] = true;
      });
    });
  });
  return active;
};

// Check if all state are active states
const areAllActiveState = (data: Record<string, boolean>): boolean => {
  const key = (Object.keys(data) as Array<string>).find((key) => data[key] === false);
  if (key) {
    return false;
  }
  return true;
};

/**
 * Card component that accepts a nested data structure (parent rows containing child rows) to display
 * StatePeriod data on a StateOverTimeChart.
 *
 * Includes legends, and logic for enabling zoom on the StateOverTimeChart.
 *
 * This gives us a fully featured component to  display state over time data,
 * with legends, that can be reused easily.
 */
const StateOverTimeCard = ({
  borderColor,
  nestedData,
  title,
  stateColors,
  sync,
  brush,
  trackingLines,
  scatterChartData,
  barSpacing,
  barCornerRadius,
  intervalSpacing,
  tickLabelPadding,
  chartPadding,
  subHeadingComponent,
  hideStateCodes,
  hideSubStepIds,
  isDefaultExpanded
}: Props): JSX.Element => {
  const [expandedRows, setExpandedRows] = useState(
    initExpandedState(nestedData, isDefaultExpanded)
  );
  const [activeState, setActiveState] = useState(initActiveState(nestedData));
  const { t } = useTranslation(['mh']);

  // Not zoomed by default
  const [zoomedDomain, setZoomedDomain] = useState<ZoomObjectTuples | undefined>(undefined);
  // True - when an area is being selecting
  const [zooming, setZooming] = useState(false);

  const legends = useMemo(() => {
    const map: Record<string, { stateName: string; stateColor: string }> = {};
    nestedData.forEach((parentRow) => {
      parentRow.data.forEach(({ stateCode, stateName }) => {
        map[stateCode] = {
          stateName: hideStateCodes ? stateName : `${stateName} (${stateCode})`,
          stateColor: stateColors[stateCode]
        };
      });

      parentRow.children?.forEach((childRow) => {
        childRow.data.forEach(({ stateCode, stateName }) => {
          map[stateCode] = {
            stateName: hideStateCodes ? stateName : `${stateName} (${stateCode})`,
            stateColor: stateColors[stateCode]
          };
        });
      });
    });
    return map;
  }, [stateColors, nestedData]);

  // Define rows for the StateOverTimeChart
  const rows: Row[] = useMemo(() => {
    const rows: Row[] = [];

    // Add the top level button rows
    nestedData.forEach((parentRow) => {
      const topLevelRow = {
        parentRowProperty: parentRow.parentProperty,
        label: parentRow.label,
        isButton: true,
        isLabel: parentRow.isLabel,
        isExpanded: expandedRows[parentRow.parentProperty],
        state: parentRow.parentProperty,
        bars: createBars(parentRow.data, stateColors, activeState, parentRow.parentProperty)
      } as Row;

      rows.push(topLevelRow);

      // Add the child rows, if this row is expanded
      if (expandedRows[parentRow.parentProperty]) {
        const nameMap: CodeNameMap = {};

        // Add mid level rows as non-expandable rows, similar to top level rows
        if ((parentRow?.children?.length || 0) > 0) {
          parentRow?.children?.forEach((child) => {
            const midLevelRow = {
              parentRowProperty: child.parentProperty,
              label: child.label,
              isButton: false,
              isLabel: child.isLabel,
              isExpanded: expandedRows[child.parentProperty],
              state: child.parentProperty,
              bars: createBars(child.data, stateColors, activeState, child.parentProperty)
            } as Row;

            rows.push(midLevelRow);

            child.data.forEach(({ stateCode, stateName }) => {
              // Combine the parent property with the child state code to make it a unique state
              const uniqueStateCode = `${child.parentProperty}-${stateCode}`;
              nameMap[uniqueStateCode] = {
                stateName,
                stateCode,
                uniqueStateCode
              };
            });
          });
          // If there are no mid level rows, add the child rows
        } else {
          parentRow.data.forEach(({ stateCode, stateName }) => {
            // Combine the parent property with the child state code to make it a unique state
            const uniqueStateCode = `${parentRow.parentProperty}-${stateCode}`;
            nameMap[uniqueStateCode] = {
              stateName,
              stateCode,
              uniqueStateCode
            };
          });

          // Then dynamically add a child row for each state
          Object.values(nameMap).forEach((val) => {
            rows.push({
              bars: createBars(
                parentRow.data.filter((bar) => bar.stateCode === val.stateCode),
                stateColors,
                activeState,
                val.uniqueStateCode
              ),
              label: hideStateCodes
                ? val.stateName
                : `${nameMap[val.uniqueStateCode].stateName} (${
                    nameMap[val.uniqueStateCode].stateCode
                  })`,
              state: val.uniqueStateCode,
              stateNameLabel: val.stateName
            } as Row);
          });
        }
      }
    });

    return rows;
  }, [nestedData, expandedRows, activeState]);

  // Sync zoom if Applicable
  useEffect(() => {
    if (sync) {
      setZoomedDomain(brush?.zoomedDomain);
      setZooming(false);
    }
  }, [brush?.zoomedDomain]);

  return (
    <Container borderColor={borderColor}>
      <Legends>
        <div>
          {typeof title === 'string' ? (
            <Typography weight="medium" size="0.8125rem">
              {title}
            </Typography>
          ) : (
            title
          )}
          {subHeadingComponent}
          {zoomedDomain?.x && (
            <ResetZoom>
              <ActionButton
                onClick={() => {
                  if (sync) brush?.resetZoom();
                  else setZoomedDomain(undefined);
                }}
                hideArrow
              >
                {t('reset_zoom')}
              </ActionButton>
            </ResetZoom>
          )}
        </div>
        <div>
          {Object.keys(legends).map((key) => (
            <LegendRow key={key}>
              <GraphLegend
                active={activeState[key]}
                onClick={() => {
                  if (areAllActiveState(activeState)) {
                    // set [key] to true, all others to false
                    setActiveState({
                      ...Object.keys(activeState).reduce((acc, curr) => {
                        return {
                          ...acc,
                          [curr]: false
                        };
                      }, {}),
                      [key]: true
                    });
                  } else {
                    if (activeState[key]) {
                      // set [key] to true, all others to false
                      setActiveState({
                        ...Object.keys(activeState).reduce((acc, curr) => {
                          return {
                            ...acc,
                            [curr]: true
                          };
                        }, {}),
                        [key]: true
                      });
                    } else {
                      // set [key] to true, keep all others current values
                      setActiveState({
                        ...activeState,
                        [key]: true
                      });
                    }
                  }
                }}
                id={key}
                label={legends[key].stateName}
                color={legends[key].stateColor}
              />
            </LegendRow>
          ))}
        </div>
      </Legends>
      <StateOverTimeChart
        padding={chartPadding ? chartPadding : graphPadding}
        tickLabelPadding={tickLabelPadding}
        rows={rows}
        onLabelClick={({ parentRowProperty }) => {
          parentRowProperty &&
            setExpandedRows({
              ...expandedRows,
              [parentRowProperty]: !expandedRows[parentRowProperty]
            });
        }}
        trackingLines={trackingLines}
        scatterChartData={scatterChartData}
        barSpacing={barSpacing}
        barCornerRadius={barCornerRadius}
        intervalSpacing={intervalSpacing}
        hasZoom={true}
        zooming={zooming}
        brush={{
          onBrushDomainChange: () => !zooming && setZooming(true),
          onBrushDomainChangeEnd: (d) => {
            if (sync) {
              brush?.onBrushDomainChangeEnd(d);
            } else {
              setZoomedDomain(d);
              setZooming(false);
            }
          },
          resetZoom: () => {
            if (sync) brush?.resetZoom();
            else setZoomedDomain(undefined);
          },
          zoomedDomain
        }}
        syncAxis={sync}
        hideSubStepIds={hideSubStepIds}
      />
    </Container>
  );
};

export default StateOverTimeCard;
