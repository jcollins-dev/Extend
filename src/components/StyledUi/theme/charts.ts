export const charts = {
  alarms: {
    'Product Alarm': `rgb(32, 36, 162)`,
    Warning: `rgb(255, 178, 102)`,
    Critical: `rgb(255, 74, 74)`
  },
  avureAlarms: {
    'B-Alarm': `rgb(32, 36, 162)`,
    'C-Alarm': `rgb(255, 178, 102)`,
    'A-Alarm': `rgb(255, 74, 74)`
  },
  dsiAlarms: {
    'B-Alarm': `rgb(32, 36, 162)`,
    'C-Alarm': `rgb(255, 178, 102)`,
    'A-Alarm': `rgb(255, 74, 74)`
  },
  proteinAlarms: {
    'Critical Alarm': 'rgb(239, 68, 68)',
    Alarm: 'rgb(166, 162, 210)',
    'Warning Information': 'rgb(95, 92, 101)',
    '': 'rgb(229, 227, 227)',
    undefined: 'rgb(229, 227, 227)'
  }
};
