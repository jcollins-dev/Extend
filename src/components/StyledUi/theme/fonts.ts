export const font = {
  res: {
    sm: `clamp(9px,.7vw,11px)`,
    base: `clamp(10px,.8vw,14px)`
  }
};
