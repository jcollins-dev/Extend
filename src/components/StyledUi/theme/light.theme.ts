const colors = {
  main: '#303E47',
  mainLight: '#556c7d',
  neg: 'white',
  primary: '#303E47',
  primaryAlt: '#303E47',
  secondaryLight: `#F1F7FF`,
  secondary: '#0A70FF',
  secondaryAlt: '#0A70FF',
  darken: 'rgba(0,0,0,.25)',
  dark: 'rgba(0,0,0,.5)',
  darker: 'rgba(0,0,0,.65)',
  darkest: 'rgba(0,0,0,.8)',
  gray: `#828285`
};

const text = {
  ...colors
};

const border = {
  main: '#C2C2C6',
  light: '#D0D5DD'
};

const bg = {
  negLightest: '#F9FAFB',
  negLighter: '#F1F3F4'
};

const headline = {
  primary: colors.primary
};

const status = {
  error: {
    lighter: '#FFEFEC',
    light: '#FFBCB2',
    base: '#EF4444'
  },
  warning: {
    lighter: '#FFEFEC',
    light: '#FFC9A2',
    base: '#EA7600'
  },
  success: {
    lighter: '#EDF8F2',
    light: '#B5E4CB',
    base: '#28B981'
  }
};

export default { text, border, bg, headline, status, ...colors };
