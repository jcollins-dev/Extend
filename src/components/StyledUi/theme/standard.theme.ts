const base = 1;
const unit = 'rem';

/** used for margins and padding */
export const space = {
  xs: `${base * 0.5}${unit}`,
  sm: `${base * 0.65}${unit}`,
  md: `${base * 0.8}${unit}`,
  base: `${base}${unit}`,
  lg: `${base * 1.25}${unit}`,
  xl: `${base * 1.4}${unit}`,
  xxl: `${base * 1.7}${unit}`
};

const fntbase = 1;
/** short-hand for font-size */
export const fnt = {
  sm: `${fntbase * 0.7}${unit}`,
  md: `${fntbase * 0.85}${unit}`,
  base: `${fntbase}${unit}`,
  lg: `${fntbase * 1.25}${unit}`,
  xl: `${fntbase * 1.5}${unit}`
};

export const radius = {
  base: '0.625rem',
  sm: '.3rem',
  xs: '5px'
};
