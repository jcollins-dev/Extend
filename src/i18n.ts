import i18next from 'i18next';

import { initReactI18next } from 'react-i18next';
import LanguageDetector from 'i18next-browser-languagedetector';
import CrowdinBackend from './i18nCrowdinOTA';

//Import all translation files
import COMMON_EN from 'locale/en/common.json';
import FPNS_EN from 'locale/en/fpns.json';

import MH_EN from 'locale/en/mh.json';
import COMMON_DE from 'locale/de/common.json';
import FPNS_DE from 'locale/de/fpns.json';
import MH_DE from 'locale/de/mh.json';
import COMMON_IT from 'locale/it/common.json';
import FPNS_IT from 'locale/it/fpns.json';
import MH_IT from 'locale/it/mh.json';

// Always include english in resources so we
// pull english strings from our local /src/locale/en files
// eslint-disable-next-line
const resources: any = {
  en: {
    common: COMMON_EN,
    fpns: FPNS_EN,
    mh: MH_EN
  }
};

if (process.env.REACT_APP_CROWDIN_INTEGRATION === 'true') {
  i18next.use(new CrowdinBackend());
} else {
  resources['de'] = {
    common: COMMON_DE,
    fpns: FPNS_DE,
    mh: MH_DE
  };

  resources['it'] = {
    common: COMMON_IT,
    fpns: FPNS_IT,
    mh: MH_IT
  };
}

i18next
  .use(LanguageDetector)
  .use(initReactI18next)
  .init({
    resources: resources,
    fallbackLng: 'en-US',
    ns: ['common', 'fpns', 'mh'],
    partialBundledLanguages: true
  });

export default i18next;
