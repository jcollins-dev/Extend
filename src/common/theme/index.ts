import * as base from './standard.theme';
import { font } from './fonts';
import light from './light.theme';
import * as effects from './effects';
import * as charts from './charts';

export const styledTheme = { effects, colors: light, font, ...charts, ...base };
