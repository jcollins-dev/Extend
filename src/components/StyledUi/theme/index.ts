import * as base from './standard.theme';
import { font } from './fonts';
import light from './light.theme';
import * as effects from './effects';
import * as charts from './charts';

export const styledTheme = { effects, color: light, font, ...charts, ...base };
