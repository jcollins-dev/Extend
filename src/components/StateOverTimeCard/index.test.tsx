import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { I18nextProvider } from 'react-i18next';
import i18n from 'i18nForTests';
import theme from 'themes';
import StateOverTimeCard from '.';

describe('StateOverTimeCard', () => {
  it('renders correctly when there are no bars', () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <I18nextProvider i18n={i18n}>
          <StateOverTimeCard title={'test'} nestedData={[]} stateColors={{}} />
        </I18nextProvider>
      </ThemeProvider>
    );
  });
});
