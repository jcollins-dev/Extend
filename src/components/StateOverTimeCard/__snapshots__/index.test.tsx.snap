// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StateOverTimeCard renders correctly when there are no bars 1`] = `
.c0 {
  overflow: hidden;
  border: 1px solid #E5E9ED;
  border-radius: 0.5rem;
  background-color: #FFFFFF;
}

.c3 {
  font-family: "SF Pro Display",-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Oxygen-Sans,Ubuntu,Cantarell,"Helvetica Neue",sans-serif;
  font-size: 0.8125rem;
  font-weight: 500;
  margin-top: 0;
  margin-bottom: 1rem;
}

.c6 {
  font-family: "SF Pro Display",-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Oxygen-Sans,Ubuntu,Cantarell,"Helvetica Neue",sans-serif;
  color: #828285;
  font-size: 0.875rem;
  font-weight: 400;
  margin-top: 0;
  margin-bottom: 0;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  height: 100%;
  text-align: right;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c7 {
  display: -webkit-box;
  -webkit-line-clamp: 2;
  -webkit-box-orient: vertical;
  white-space: normal;
  overflow: hidden;
  line-height: 1.2;
  font-size: 0.8125rem;
}

.c4 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c2 {
  width: 11.5625rem;
  padding: 1rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  background-color: #F9FAFB;
  border-right: 1px solid #F1F3F4;
}

<div>
  <div
    class="c0 c1"
  >
    <div
      class="c2"
    >
      <div>
        <p
          class="c3"
        >
          test
        </p>
      </div>
      <div />
    </div>
    <div
      class="c4"
    >
      <div
        class="VictoryContainer"
        style="height: auto; pointer-events: none; position: relative; width: 100%;"
      >
        <svg
          height="50"
          role="img"
          style="pointer-events: all; width: 100%; height: 100%;"
          viewBox="0 0 0 50"
          width="0"
        >
          <g
            role="presentation"
          >
            <line
              role="presentation"
              shape-rendering="auto"
              style="stroke: none; fill: transparent; stroke-width: 1; stroke-linecap: round; stroke-linejoin: round;"
              vector-effect="non-scaling-stroke"
              x1="150"
              x2="150"
              y1="50"
              y2="45"
            />
            <g
              role="presentation"
            >
              <g
                transform="translate(0,45.65832784726794)"
              >
                <foreignobject
                  height="32.5"
                  width="140"
                  x="0"
                  y="-16.25"
                >
                  <div
                    class="c5"
                  >
                    <span
                      class="c6"
                      color="mediumGrey3"
                    >
                      <span
                        class="c7"
                      >
                        undefined  
                      </span>
                    </span>
                  </div>
                </foreignobject>
              </g>
            </g>
            <g
              role="presentation"
            >
              <g
                transform="translate(0,46.316655694535875)"
              >
                <foreignobject
                  height="32.5"
                  width="140"
                  x="0"
                  y="-16.25"
                >
                  <div
                    class="c5"
                  >
                    <span
                      class="c6"
                      color="mediumGrey3"
                    >
                      <span
                        class="c7"
                      >
                        undefined  
                      </span>
                    </span>
                  </div>
                </foreignobject>
              </g>
            </g>
            <g
              role="presentation"
            >
              <g
                transform="translate(0,46.97498354180382)"
              >
                <foreignobject
                  height="32.5"
                  width="140"
                  x="0"
                  y="-16.25"
                >
                  <div
                    class="c5"
                  >
                    <span
                      class="c6"
                      color="mediumGrey3"
                    >
                      <span
                        class="c7"
                      >
                        undefined  
                      </span>
                    </span>
                  </div>
                </foreignobject>
              </g>
            </g>
            <g
              role="presentation"
            >
              <g
                transform="translate(0,47.63331138907176)"
              >
                <foreignobject
                  height="32.5"
                  width="140"
                  x="0"
                  y="-16.25"
                >
                  <div
                    class="c5"
                  >
                    <span
                      class="c6"
                      color="mediumGrey3"
                    >
                      <span
                        class="c7"
                      >
                        undefined  
                      </span>
                    </span>
                  </div>
                </foreignobject>
              </g>
            </g>
            <g
              role="presentation"
            >
              <g
                transform="translate(0,48.29163923633969)"
              >
                <foreignobject
                  height="32.5"
                  width="140"
                  x="0"
                  y="-16.25"
                >
                  <div
                    class="c5"
                  >
                    <span
                      class="c6"
                      color="mediumGrey3"
                    >
                      <span
                        class="c7"
                      >
                        undefined  
                      </span>
                    </span>
                  </div>
                </foreignobject>
              </g>
            </g>
            <g
              role="presentation"
            >
              <g
                transform="translate(0,48.94996708360764)"
              >
                <foreignobject
                  height="32.5"
                  width="140"
                  x="0"
                  y="-16.25"
                >
                  <div
                    class="c5"
                  >
                    <span
                      class="c6"
                      color="mediumGrey3"
                    >
                      <span
                        class="c7"
                      >
                        undefined  
                      </span>
                    </span>
                  </div>
                </foreignobject>
              </g>
            </g>
            <g
              role="presentation"
            >
              <g
                transform="translate(0,49.60829493087558)"
              >
                <foreignobject
                  height="32.5"
                  width="140"
                  x="0"
                  y="-16.25"
                >
                  <div
                    class="c5"
                  >
                    <span
                      class="c6"
                      color="mediumGrey3"
                    >
                      <span
                        class="c7"
                      >
                        undefined  
                      </span>
                    </span>
                  </div>
                </foreignobject>
              </g>
            </g>
          </g>
          <g
            role="presentation"
          >
            <line
              role="presentation"
              shape-rendering="auto"
              style="stroke: none; fill: transparent; stroke-width: 1; stroke-linecap: round; stroke-linejoin: round;"
              vector-effect="non-scaling-stroke"
              x1="150"
              x2="-30"
              y1="50"
              y2="50"
            />
            <g
              role="presentation"
            >
              <line
                role="presentation"
                shape-rendering="auto"
                style="stroke: #E5E9ED; fill: none; pointer-events: painted;"
                vector-effect="non-scaling-stroke"
                x1="150"
                x2="150"
                y1="50"
                y2="45"
              />
              <text
                direction="inherit"
                dx="0"
                id="chart-undefined-1-tickLabels-0"
                x="150"
                y="25.259999999999998"
              >
                <tspan
                  dx="0"
                  dy="0"
                  style="padding: 10px; font-family: \\"SF Pro Display\\", -apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, Oxygen-Sans, Ubuntu, Cantarell, \\"Helvetica Neue\\", sans-serif; font-size: 12px; letter-spacing: normal; fill: #828285; stroke: transparent;"
                  text-anchor="middle"
                  x="150"
                >
                  12:00 AM
                </tspan>
                <tspan
                  dx="0"
                  dy="12"
                  style="padding: 10px; font-family: \\"SF Pro Display\\", -apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, Oxygen-Sans, Ubuntu, Cantarell, \\"Helvetica Neue\\", sans-serif; font-size: 12px; letter-spacing: normal; fill: #828285; stroke: transparent;"
                  text-anchor="middle"
                  x="150"
                >
                  01/01
                </tspan>
              </text>
            </g>
            <g
              role="presentation"
            >
              <line
                role="presentation"
                shape-rendering="auto"
                style="stroke: #E5E9ED; fill: none; pointer-events: painted;"
                vector-effect="non-scaling-stroke"
                x1="-30"
                x2="-30"
                y1="50"
                y2="45"
              />
              <text
                direction="inherit"
                dx="0"
                id="chart-undefined-1-tickLabels-1"
                x="-30"
                y="25.259999999999998"
              >
                <tspan
                  dx="0"
                  dy="0"
                  style="padding: 10px; font-family: \\"SF Pro Display\\", -apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, Oxygen-Sans, Ubuntu, Cantarell, \\"Helvetica Neue\\", sans-serif; font-size: 12px; letter-spacing: normal; fill: #828285; stroke: transparent;"
                  text-anchor="middle"
                  x="-30"
                >
                  12:00 AM
                </tspan>
                <tspan
                  dx="0"
                  dy="12"
                  style="padding: 10px; font-family: \\"SF Pro Display\\", -apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Roboto, Oxygen-Sans, Ubuntu, Cantarell, \\"Helvetica Neue\\", sans-serif; font-size: 12px; letter-spacing: normal; fill: #828285; stroke: transparent;"
                  text-anchor="middle"
                  x="-30"
                >
                  01/01
                </tspan>
              </text>
            </g>
          </g>
          <g
            role="presentation"
          />
          <rect
            cursor="move"
            height="5"
            style="stroke: transparent; fill: black; fill-opacity: 0;"
            vector-effect="non-scaling-stroke"
            width="180"
            x="-30"
            y="45"
          />
          <rect
            cursor="ew-resize"
            height="5"
            style="stroke: transparent; fill: transparent;"
            vector-effect="non-scaling-stroke"
            width="8"
            x="-34"
            y="45"
          />
          <rect
            cursor="ew-resize"
            height="5"
            style="stroke: transparent; fill: transparent;"
            vector-effect="non-scaling-stroke"
            width="8"
            x="146"
            y="45"
          />
        </svg>
        <div
          style="z-index: 99; position: absolute; top: 0px; left: 0px; width: 100%; height: 100%;"
        >
          <svg
            height="50"
            style="overflow: visible; width: 100%; height: 100%;"
            viewBox="0 0 0 50"
            width="0"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
